File: py/common/sequences.py
#!/usr/bin/env python3

"""sequences.py

Functions for producing and operating on numerical sequences.
"""

__author__ = 'Curtis Belmonte'

import functools
import itertools
import math
import operator
from typing import (
    Callable,
    Dict,
    Iterable,
    Iterator,
    Sequence,
    Set,
)

import common.arithmetic as arith


def _compute_chain_length(
        lengths: Dict[int, int],
        n: int,
        step: Callable[[int], int],
        is_valid: Callable[[int], bool],
        invalid_set: Set[int]) -> None:

    """Single-chain helper function for compute_chain_lengths.

    Updates lengths and invalid_set based on the sequence starting from n.
    """

    # initialize map of terms to their positions in sequence
    terms = {} # type: Dict[int, int]

    while True:
        # if chain is invalid, mark all terms accordingly
        if not is_valid(n) or n in invalid_set or n in lengths:
            for n in terms:
                invalid_set.add(n)
            return

        # if completed chain, set length for all terms in it
        elif n in terms:
            index = terms[n]
            length = len(terms) - index
            for n, i in terms.items():
                invalid_set.add(n)
                if i >= index:
                    lengths[n] = length
            return

        # otherwise, continue building the current chain
        else:
            terms[n] = len(terms)

        # advance to next number in chain
        n = step(n)


def arithmetic_product(a: int, n: int, d: int = 1) -> int:
    """Returns the product of the arithmetic sequence with parameters a, n, d.

    a: The first term in the sequence
    n: The total number of terms in the sequence
    d: The difference between any two terms in the sequence
    """

    return functools.reduce(operator.mul, range(a, a + n * d, d), 1)


def arithmetic_series(a: int, n: int, d: int = 1) -> int:
    """Returns the sum of the arithmetic sequence with parameters a, n, d.

    a: The first term in the sequence
    n: The total number of terms in the sequence
    d: The difference between any two terms in the sequence
    """

    return n * (2 * a + (n - 1) * d) // 2


def collatz_step(n: int) -> int:
    """Returns the next number in the Collatz sequence following n."""
    return n // 2 if n % 2 == 0 else 3 * n + 1


def compute_chain_lengths(
        lengths: Dict[int, int],
        values: Iterable[int],
        step: Callable[[int], int],
        is_valid: Callable[[int], bool] = lambda x: True) -> None:

    """Populates lengths with chain lengths starting from each term in values.

    lengths: The dict to be populated, mapping each term to its chain length
    values: An iterable of all valid starting terms for a chain
    step: For any term n, step(n) gives the next term in the chain
    is_valid: is_valid(n) should return True iff n is a valid chain member
    """

    invalid_set = set() # type: Set[int]
    for n in values:
        _compute_chain_length(lengths, n, step, is_valid, invalid_set)


def fibonacci(n: int) -> int:
    """Returns the nth Fibonacci number, with F(0) = F(1) = 1."""
    return next(itertools.islice(generate_fibonacci(), n, n + 1))


def generate_fibonacci() -> Iterator[int]:
    """Yields each number in the Fibonacci sequence, beginning with 1, 1."""
    fib_prev = 0
    fib_curr = 1
    while True:
        yield fib_curr
        fib_prev, fib_curr = fib_curr, fib_prev + fib_curr


def generate_products(factors: Sequence[int], cache_capacity: int = 1000)\
        -> Iterator[int]:

    """Yields all distinct products of powers of factors in increasing order.

    Generates an infinite sequence of integers, starting with 1. If provided,
    cache_capacity determines the maximum size of the internal list used for
    caching previous products.

    Adapted from: https://rosettacode.org/wiki/Hamming_numbers#Python
    """

    # initialize product, cache, and next multiples
    product = 1
    product_cache = [product]
    indices = [0] * len(factors)
    multiples = [x * product_cache[i] for x, i in zip(factors, indices)]
    yield product

    while True:
        # poll next product and update list of next multiples
        product = min(multiples)
        product_cache.append(product)
        for i, (m, d, j) in enumerate(zip(multiples, factors, indices)):
            if m == product:
                j += 1
                indices[i] = j
                multiples[i] = d * product_cache[j]

        # trim the cache if it's over capacity
        min_index = min(indices)
        if min_index >= cache_capacity:
            del product_cache[:min_index]
            indices = [i - min_index for i in indices]

        yield product


def hexagonal(n: int) -> int:
    """Returns the nth hexagonal number, starting with 1."""
    return n * (2 * n - 1)


def is_fibonacci(n: int) -> bool:
    """Determines if the natural number n is a Fibonacci number."""
    if n < 2**52:
        # use arithmetic test if within floating-point precision
        term = 5 * n**2
        return is_square(term + 4) or is_square(term - 4)
    else:
        # generate Fibonacci numbers until n is found or exceeded
        for fib_num in generate_fibonacci():
            if fib_num >= n:
                return fib_num == n


def is_hexagonal(n: int) -> bool:
    """Determines if the natural number n is a hexagonal number."""
    radical_sum = 1 + (8 * n)
    return is_square(radical_sum) and arith.int_sqrt(radical_sum) % 4 == 3


def is_pentagonal(n: int) -> bool:
    """Determines if the natural number n is a pentagonal number."""
    radical_sum = 1 + (24 * n)
    return is_square(radical_sum) and arith.int_sqrt(radical_sum) % 6 == 5


def is_power(n: int, p: int) -> bool:
    """Determines if the natural number n is a perfect power with exponent p.

    Specifically, returns True iff n = m^p for some natural number m.
    """

    root_n = n**(1 / p)
    lo_power = (int(root_n))**p
    hi_power = (int(math.ceil(root_n)))**p
    return lo_power == n or hi_power == n


def is_square(n: int) -> bool:
    """Determines if the natural number n is a perfect square."""
    return (int(math.sqrt(n)))**2 == n


def is_triangular(n: int) -> bool:
    """Determines if the natural number n is a triangle number."""
    return is_square(8 * n + 1)


def next_multiple(n: int, min_val: int) -> int:
    """Returns the next multiple of the natural number n >= min_val."""
    return min_val + ((n - (min_val % n)) % n)


def pentagonal(n: int) -> int:
    """Returns the nth pentagonal number, starting with 1."""
    return n * (3 * n - 1) // 2


def sum_of_squares(n: int) -> int:
    """Returns the sum of the squares of the first n natural numbers."""
    return (2 * n**3 + 3 * n**2 + n) // 6


def triangular(n: int) -> int:
    """Returns the nth triangle number, starting with 1.

    The resulting value is also the sum of the first n natural numbers.
    """

    return n * (n + 1) // 2

File: jl/src/Sequences.jl
#!/usr/bin/env julia

"""
Utilities for producing and operating on numerical sequences.
"""
module Sequences

"""
    Fibonacci([T::Type{<:Integer}])

Iterable object that returns subsequent terms of the Fibonacci sequence.

## Fields
- `T`: The integer type of each term in the sequence. Defaults to `Int`

## Examples
```julia-repl
julia> first(Fibonacci())
1

julia> first(Fibonacci(UInt8))
0x01

julia> for n in Fibonacci()
           println(n)
           if n > 10
               break
           end
       end
1
2
3
5
8
13

julia> for (i, n) in Iterators.enumerate(Fibonacci())
           println(n)
           if i == 5
               break
           end
       end
1
2
3
5
8
```

See also: [`fibonacci`](@ref)
"""
struct Fibonacci
    T::Type{<:Integer}
end
Fibonacci() = Fibonacci(Int)

function Base.iterate(
    iter::Fibonacci,
    state::Tuple{Integer, Integer} = (zero(iter.T), oneunit(iter.T))
)
    fib_old, fib_new = state
    fib_old, fib_new = fib_new, (fib_old + fib_new)
    return fib_new, (fib_old, fib_new)
end

"""
    arithmetic_series(a::Integer, n::Integer, d::Integer = 1)

Returns the sum of the arithmetic sequence with parameters `a`, `n`, `d`.

# Arguments
- `a`: The first term in the sequence
- `n`: The total number of terms in the sequence
- `d`: The difference between two consecutive terms in the sequence

# Preconditions
- `n ≥ 0`

# Examples
```jldoctest
julia> arithmetic_series(1, 4)
10

julia> arithmetic_series(2, 6, -1)
-3

julia> arithmetic_series(3, 5, 4)
55
```
"""
function arithmetic_series(a::Integer, n::Integer, d::Integer = 1)
    return n * (2a + (n - 1)d) ÷ 2
end

"""
    fibonacci([T::Type{<:Integer}], n::Integer)

Computes the specified term `F(n)` of the Fibonacci sequence, defined by the
relation `F(m) = F(m - 1) + F(m - 2)` with `F(0) = F(1) = 1`.

## Arguments
- `T`: The integer type of each term in the sequence. Defaults to `typeof(n)`
- `n`: Specifies the term `F(n)` in the Fibonacci sequence to return

## Preconditions
- `n ≥ 0`

## Examples
```jldoctest
julia> fibonacci(0)
1

julia> fibonacci(4)
5

julia> fibonacci(UInt8, 5)
0x08
```

See also: [`Fibonacci`](@ref)
"""
function fibonacci(n::Integer)
    return fibonacci(typeof(n), n)
end
function fibonacci(T::Type{<:Integer}, n::Integer)
    # Fibonacci iterates from F(1), so handle F(0) ourselves
    if n == 0
        return oneunit(T)
    end

    for (i, fib_num) in enumerate(Fibonacci(T))
        if i == n
            return fib_num
        end
    end

    # Fibonacci iterator should never terminate
    throw(ErrorException("iteration terminated unexpectedly"))
end

"""
    sum_of_squares(n::Integer)

Computes the sum of the squares of the first `n` natural numbers.

## Preconditions
- `n ≥ 0`

## Examples
```jldoctest
julia> sum_of_squares(1)
1

julia> sum_of_squares(0x02)
0x05

julia> sum_of_squares(4)
30
```
"""
function sum_of_squares(n::Integer)
    T = typeof(n)
    cube_term = convert(T, 2) * n^3
    square_term = convert(T, 3) * n^2
    return (cube_term + square_term + n) ÷ convert(T, 6)
end

"""
    triangular(n::Integer)

Computes the `n`th triangular number `T(n)`, with `T(1) = 1`. The resulting
value is also the sum of the first `n` natural numbers.

## Preconditions
- `n ≥ 0`

## Examples
```jldoctest
julia> triangular(1)
1

julia> triangular(0x03)
0x06

julia> triangular(4)
10
```
"""
triangular(n::Integer) = n * (n + oneunit(n)) ÷ convert(typeof(n), 2)

end
